### PR TITLE
add remark plugin to replace asides

### LINKED ISSUE #967: Integration page asides
Here’s something that I didn’t get to while working on #954 that would be cool to add to the integration READMEs workflow.

The READMEs currently use a mix of styles for tips and warnings using blockquotes as is common on GitHub. I would like to see these standardised to [GitHub’s new syntax for notes and warnings](https://github.com/github-community/community/discussions/16925):

```md
> **Note**
> This is a note

> **Warning**
> This is a warning
```

> **Note**
> This is a note

> **Warning**
> This is a warning

**THEN** I would like to add a remark plugin to our processing pipeline here:

https://github.com/withastro/docs/blob/2c93d8ff0615ac11a6ddcc0b009636c166530fe0/scripts/generate-integration-pages.ts#L73-L77

That finds that syntax and converts it to:

```md
:::note
This is a note
:::

:::caution
This is a warning
:::
```

### PR DESCRIPTION
**Pull request by kevinzunigacuellar**
#### What kind of changes does this PR include?

- Changes to the docs site code

#### Description

- Closes #967 
- Probably needs a lot of refactoring but the main idea is there 😇

### CODE REVIEW

File: scripts/generate-integration-pages.ts [hunk 1]
```diff
@@ -178,6 +179,47 @@ function absoluteLinks({ base }: { base: string }) {
 	};
 }
 
+/** Remark plugin to replace GitHub . */
+function replaceAsides() {
+	return function transform(tree: Root) {
+		// visit blockquote
+		visit(tree, 'blockquote', (node) => {
+			// first child (<blockquote><p>...</p></blockquote>)
+			const [firstChild, trailingText] = node.children[0].children;
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:188]: It can be a bit hard to follow all the nested `children` in remark plug-ins, so I personally find it can help declaring variables so we get some more explicit names, like:

```suggestion
			const openingParagraph = node.children[0];
			const [firstChild, trailingText] = openingParagraph.children;
```

File: scripts/generate-integration-pages.ts [hunk 2]
```diff
@@ -178,6 +179,47 @@ function absoluteLinks({ base }: { base: string }) {
...
+
+			if (firstChild.type === 'strong') {
+				if (
+					// check for **Note** or **Warning**
+					firstChild.children[0].value.includes('Note') ||
+					firstChild.children[0].value.includes('Warning')
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:194]: I might be tempted to reach for a regular expression here:

```suggestion
					/Note|Warning/.test(firstChild.children[0].value)
```

File: scripts/generate-integration-pages.ts [hunk 3]
```diff
@@ -178,6 +179,47 @@ function absoluteLinks({ base }: { base: string }) {
...
+				) {
+					const AsideType =
+						firstChild.children[0].value.toLowerCase() === 'note' ? 'note' : 'caution';
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:197]: This technically makes “caution” our default note type. Not sure we’d ever actually hit this, but just in case I think defaulting to “note” probably makes more sense:

```suggestion
						firstChild.children[0].value.toLowerCase() === 'warning' ? 'caution' : 'note';
```

File: scripts/generate-integration-pages.ts [hunk 4]
```diff
@@ -178,6 +179,47 @@ function absoluteLinks({ base }: { base: string }) {
...
+
+					// remove blockquotes `>`
+					node.type = 'paragraph';
+
+					// if starts with `: ` remove it
+					if (trailingText.value.startsWith(': ')) {
+						trailingText.value = trailingText.value.replace(/^: /, '\n');
+					}
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:205]: Should be safe to skip the `if` here I think: if `/^: /` doesn’t match, nothing will be replaced anyway.

```suggestion
					trailingText.value = trailingText.value.replace(/^: /, '\n');
```

Comment (delucis) [reply scripts/generate-integration-pages.ts:205]: Question: do we need to add the additional `\n`? If so, I’d love the comment in the line above to mention why.

**Review by delucis — COMMENTED** [review]
This is awesome — so happy we (YOU!) finally did this!

Left some comments, but seems pretty great to me 😀

File: scripts/generate-integration-pages.ts (hunk 3 above)
Comment (kevinzunigacuellar) [reply scripts/generate-integration-pages.ts:197]: caution this may explode CI 😆🔥😇

**Comment by kevinzunigacuellar** [comment]
'we are Groot' 💜

File: scripts/generate-integration-pages.ts [hunk 5]
```diff
@@ -178,6 +179,41 @@ function absoluteLinks({ base }: { base: string }) {
 	};
 }
 
+/** Remark plugin to replace GitHub . */
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:182]: Nit: this comment is complete!

```suggestion
/** Remark plugin to replace GitHub note/warning syntax with docs-style asides. */
```

File: scripts/generate-integration-pages.ts [hunk 6]
```diff
@@ -178,6 +179,41 @@ function absoluteLinks({ base }: { base: string }) {
...
+function replaceAsides() {
+	return function transform(tree: Root) {
+		visit(tree, 'blockquote', (node) => {
+			const openingParagraph = node.children[0];
+			const [firstChild, trailingText] = openingParagraph.children;
+
+			// check for **Note:** or **Warning:** at the beginning of the first paragraph
+			if (firstChild.type === 'strong' && /Note|Warning/.test(firstChild.children[0].value)) {
```
Comment (delucis) [inline scripts/generate-integration-pages.ts:190]: Style nit you can skip if you prefer: this could be an opportunity for an early return:

```js
if (
	firstChild.type !== 'strong') ||
	!/Note|Warning/.test(firstChild.children[0].value)
) {
	return;
}
```

That would avoid nesting everything that follows inside the `if` statement. But it’s fine as-is too.

**Review by delucis — COMMENTED** [review]
Couple of small nits here, but once you’ve accepted or dismissed these — LGTM!

File: scripts/generate-integration-pages.ts (hunk 6 above)
Comment (kevinzunigacuellar) [reply scripts/generate-integration-pages.ts:190]: I think I saw early returns as common practice in remark plugins. I will add this 😇

**Review by delucis — APPROVED** [review]
LGTM! Awesome work on this one Kevin!

Fun fact about the early returns, a lot of the remark utilities like `visit` we’re using here, secretly support special constants to control them:

```js
import { visit, CONTINUE, EXIT, SKIP } from 'unist-util-visit';

function transform(tree) {
  visit(tree, node => {
    if (node.notThisOne) return CONTINUE;
    if (node.notThisOrItsChildren) return SKIP;
    if (node.nothingAfterThis) return EXIT;
  });
}
```
